### ISSUE TITLE
perf(classify): pipeline the waveform GPU classify path — CPU prep and GPU scoring are fully serial

### ISSUE THREAD

**Issue by jayhesselberth** [rnabioco/escapepod-rs#351]
## Summary

`escpod classify --device gpu` for the `waveform_model` (windowed TCN) variant alternates between long CPU-only phases and short GPU-only bursts instead of overlapping them, so the GPU sits idle for most of a run on a CPU-constrained node.

## How this was found

Observed live on a production job (escpod 0.23.0-gpu, model `charging_tcn_sup6_rna004@v0.1.0`, `--device gpu --threads 4`, 4 CPUs / 1 GPU on a Slurm node) mid-run on `~432k` anchored reads: `nvidia-smi dmon` showed sustained 0% SM utilization for 30+ consecutive seconds while the `escpod` process sat at 341–386% CPU (of its 4-CPU allocation) and `/proc/<pid>/io`'s `read_bytes`/`syscr` were completely flat over a 25s window, with every rayon worker thread parked in `futex_wait_queue`. This first looked like a hang (and briefly triggered a false "BeeGFS RPC stall" theory), but a few minutes later the GPU burst to life and the run kept making forward progress — the flat CPU/IO snapshot was real, just not a stall.

## Root cause

`classify_reads_gpu` (`crates/escapepod-classify/src/waveform.rs:558-624`) processes reads in memory-bounded superbatches:

```rust
const SUPERBATCH: usize = 16_384;   // waveform.rs:579

for slice in reads.chunks(SUPERBATCH) {
    let prepared: Vec<Prepared> = slice
        .par_iter()                  // ALL CPU: signal fetch + banded-DP refinement
        .map(|&read| { /* signal_adc + assemble_chunk */ })
        .collect::<Result<_>>()?;
    // ^ blocks here until the WHOLE superbatch is prepared — zero GPU calls yet

    let mut groups = ready.chunks_exact(batch);   // batch = gpu.batch_size(), default 128
    for group in &mut groups {
        let logits = gpu.logits(&refs, spec)?;    // ALL GPU: ~128 sequential calls, CPU otherwise idle
        ...
    }
}
```

Per superbatch (16,384 reads): first every read's signal fetch (`signal_adc`) and chunk assembly (`assemble_chunk` — banded-DP refinement + k-mer residual) runs on the `rayon` pool with **zero** GPU calls issued, fully saturating whatever CPU allocation the job has; only once that entire superbatch is ready does the loop move to a **sequential** run of `gpu.logits()` calls (`batch = gpu.batch_size()`, default 128, override `ESCAPEPOD_WAVEFORM_GPU_BATCH`), during which the CPU does essentially nothing but await each call. There is no producer/consumer overlap between the two phases.

Contrast `escapepod-demux/src/adapter_cnn_gpu.rs::detect_prepped` (`adapter_cnn_gpu.rs:196-199`), which is explicit about the pattern this path is missing:

> "Lets callers run [`AdapterCnnConfig::prep`] in parallel on CPU producer threads and feed prepped blocks to the GPU, so the GPU thread only does grouping + inference + decode."

## Why this matters

The per-call GPU speedup for this graph is real and well measured in isolation — `Stage::WaveformTcn::cpu_cost()` (`crates/escapepod-cli/src/device.rs:200`) cites "~11x slower than GPU at production batch sizes", and `examples/tcn_cuda_probe.rs` (commit 29bde69, #337) measured 0.59 ms/read GPU vs 6.56 ms/read CPU at batch 128. But none of that per-call win reaches wall-clock if the GPU is idle for the entire time the *next* superbatch's CPU prep is running — which is exactly what a CPU-constrained node (this job: 4 cores) makes worse, since there's little spare CPU capacity even to consider overlapping the two phases today, and the code doesn't attempt to regardless of CPU headroom.

**Not the #347/#348 cudart-probe bug**: that panics at model-load time before any read is scored (`waveform_net_gpu.rs`'s `cudart_runtime_probe`), which would kill the process outright. This job placed onto GPU successfully and produced real logits — the symptom here is scheduling, not a broken runtime.

## Acceptance criteria

- CPU-side prep (`signal_adc` + `assemble_chunk`) for one superbatch overlaps with GPU-side scoring (`gpu.logits()`) of previously-prepared work — the GPU is not idle while prepared-but-unscored reads exist, and CPU prep for later reads proceeds while a GPU batch call is in flight.
- Bit-identical output to current `classify_reads_gpu`: same `ReadCall`s, same `NoCall` tallies, same sort order (`charging_waveform_chunks.rs`-style parity test, or extend it). This is a scheduling change, not a numerics change — do not touch `expand_instance_norm`, `hoist_conv_padding`, or the batch-1-is-wrong finding documented in `waveform_net_gpu.rs`.
- A new end-to-end (not isolated-kernel) measurement — e.g. `crates/escapepod-classify/examples/waveform_gpu_pipeline_probe.rs`, sibling to the existing `tcn_cuda_probe.rs`/`axpy_probe.rs` probes — showing wall-clock improvement on a real bundle + real POD5/BAM input at a CPU allocation representative of production use (~4 cores), plus a `nvidia-smi dmon` before/after duty-cycle capture in the PR description (the way `project_fused_crf_demux_gpu_starved`'s write-up did for the CRF path).
- `gpu.batch_size()` / `ESCAPEPOD_WAVEFORM_GPU_BATCH` and `SUPERBATCH`'s memory-bound rationale (chunk size × in-flight count) are preserved — this is about overlapping the two phases, not about changing batch sizing.

## Non-goals

- Not a mandate to copy `adapter_cnn_gpu.rs`'s bounded-channel implementation verbatim — that crate's shape (variable-length CNN inputs grouped by exact length, sub-batch OOM retry) differs from this path's fixed-shape chunks. Reuse the *pattern* (CPU producer thread pool → bounded channel → single GPU consumer thread), not the code.
- Does not touch the CPU-only path (`classify_reads` / `WaveformNet`, batch-1-by-design) — only `classify_reads_gpu`.
- Does not revisit `SUPERBATCH = 16_384` itself, or the memory-bound reasoning behind it (`waveform.rs:551-553`) — only how the two phases inside each superbatch are scheduled relative to each other.
- No claim that the isolated-kernel numbers in `device.rs`/`waveform_net_gpu.rs` are wrong — they're correct in isolation; this issue is specifically about the missing overlap in the surrounding pipeline.

## Environment (for reproduction)

- escpod 0.23.0-gpu (current `main` HEAD's `waveform.rs` is byte-identical to 0.23.0 in the relevant function — only `waveform_net_gpu.rs`'s load-time `cudart_runtime_probe` from #348 has changed since)
- `--device gpu --threads 4`, bundle `charging_tcn_sup6_rna004@v0.1.0`, A30 GPU
- `nvidia-smi dmon -s u -c 20 -d 1` during the prep phase of a superbatch: sustained `sm=0` for the full window

**Comment by jayhesselberth** [comment]
Shipped in bae51ae / PR #352 (`perf(classify): pipeline the waveform GPU classify path`): `classify_reads_gpu` now runs a dedicated GPU consumer thread reading prepped groups off a bounded channel while CPU producer threads (via `rayon`) prep the next `prep_chunk` of reads, overlapping CPU prep with GPU scoring instead of alternating fully-serial phases. Confirmed present in current `main` and further tuned in #361 (`groups_in_flight`/`prep_chunk` defaults re-swept for `--threads 16`). This issue was left open after #352 merged; closing now since the acceptance criteria are met.
